[RLlib] Learner API (+DreamerV3 prep): `Learner.hps` per Module (support algo config overrides on new stack), other cleanups

sven1977 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

Learner API (+DreamerV3 prep): `Learner.hps` per Module (support algo config overrides on new stack), other cleanups

This PR ...:
- Adds support for per-module AlgorithmConfig overrides (e.g. using a different learning rate for different modules).
- Gets rid of Learner.optimizer_config (in favor of moving these settings into the LearnerHyperparameters)
- Simplifies Schedule configs for e.g. lr, entropy_coeff, etc.. When using a schedule, the user sets the exact same config setting than when providing a fixed value. This avoids confusion as these two modes are exclusive (either fixed value OR schedule).
- Restricts optimizer to either one of the following two scenarios:
- - User uses config.optimizer_type="adam", lr=..., grad_clip=... and RLlib will automatically create this optimizer (per module) and handle the lr/scheduling-setup.
- - User overrides a custom Learner's `configure_optimizers_for_module` method and handle optimizer creation, configuration, AND lr scheduling all by themselves (RLlib will NOT touch the learning rate ever, no matter the config).


<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

sven1977 commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -330,8 +329,8 @@ def __init__(self, algo_class=None):
         self.gamma = 0.99
+        self.optimizer_type = "adam"
         self.lr = 0.001
-        self.lr_schedule = None
```
`lr_schedule` is no longer supported on the new stack, but still on those old-stack algos that had this as an algo-specific setting, e.g. PPO (on old stack).
For the new stack, users should simply use `lr: [ls-schedule setup]`.

sven1977 commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -3206,24 +3226,15 @@ def get_learner_group_config(self, module_spec: ModuleSpec) -> LearnerGroupConfi
-                    "grad_clip_by": self.grad_clip_by,
-                },
-                learner_hyperparameters=self.get_learner_hyperparameters(),
+                learner_hyperparameters=self.get_learner_hyperparameters(module_spec),
```
no more optimizer_config! All config information is now passed via `LearnerHyperparameters` to the Learner.

sven1977 commented on rllib/algorithms/appo/appo_learner.py:
```diff
@@ -43,14 +44,18 @@ class AppoLearner(ImpalaLearner):
+
+        # The current kl coefficients per module as tensor variables
+        # (framework specific).
+        self.curr_kl_coeffs_per_module: LambdaDefaultDict[
```
Using this new defaultdict subclass avoids us having to tediously override the `add_module` method.

sven1977 commented on rllib/algorithms/appo/appo_learner.py:
```diff
@@ -63,9 +68,10 @@ def additional_update_for_module(
         self,
         *,
         module_id: ModuleID,
+        hps: AppoLearnerHyperparameters,
```
We are passing the module specific settings now into all `.._for_module` method calls. Strictly, this is not required as users could easily get these inside their implementations by doing:

```
def additional_update_for_module(self, module_id, ...):
   hps = self.hps.get_hps_for_module(module_id)
```
Not sure, might be cleaner this way, but I'd be also happy to remove this to keep the method signatures smaller.

sven1977 commented on rllib/algorithms/impala/impala_learner.py:
```diff
@@ -50,25 +54,35 @@ class ImpalaLearner(Learner):
+            )
         )
 
+    @override(Learner)
```
This was a bug: Forgot to do this, so right now, when removing a module, its entropy_coeff Schedule remains in memory.

sven1977 commented on rllib/algorithms/tests/test_algorithm_config.py:
```diff
@@ -192,6 +194,59 @@ class A:
         config.validate()
         self.assertEqual(config.rl_module_spec.module_class, A)
 
+    def test_learner_hyperparameters_per_module(self):
```
New test case, checking the correct `algorithm_config_override` propagation into the LearnerHPs.

sven1977 commented on rllib/algorithms/tests/test_algorithm_config.py:
```diff
@@ -361,6 +336,33 @@ def get_default_rl_module_spec(self):
         spec, expected = self._get_expected_marl_spec(config, CustomRLModule1)
         self._assertEqualMARLSpecs(spec, expected)
 
+        ########################################
```
Added this sub-test to cover per-module algo config overrides.

sven1977 commented on rllib/core/learner/learner.py:
```diff
@@ -111,12 +112,51 @@ class LearnerHyperparameters:
-    # TODO (Sven): Move lr from - currently - optimizer config to only exist here.
-    # lr: float = None
+    # Parameters used for gradient postprocessing (clipping) and gradient application.
+    optimizer_type: str = None
```
4 base settings supported by RLlib for auto-constructing one optimizer per module.

If users want any other behavior:
- more than 1 optimizer per module
- custom type of optimizer with custom settings (e.g. Adam epsilon, etc..)
they must override `configure_optimizers_for_module` and then RLlib will NO longer manage the lr/lr-schedule for them. Everything has to be done manually then by the user's own Learner subclass.

sven1977 commented on rllib/core/learner/learner.py:
```diff
@@ -111,12 +112,51 @@ class LearnerHyperparameters:
+    # correct mappings from ModuleID to the .
+    _per_module_overrides: Optional[Dict[ModuleID, "LearnerHyperparameters"]] = None
+
+    def get_hps_for_module(self, module_id: ModuleID) -> "LearnerHyperparameters":
```
New API inside LearnerHPs to get module-specific HP sub-instance.

sven1977 replied:
You wouldn't call it with a module_id=None. It's a mandatory arg.

sven1977 replied:
Also, some module_ids might not want to override anything, so they'll not be found in the mapping (and just use the top-level HPs object). So no error should be thrown here.

sven1977 replied:
Leaving as-is. Reasoning:

self.hps (top-level HPs object):
    get_hps_for_module(abc) -> "I don't have overrides for abc -> return myself"
    get_hps_for_module(abc) -> "I don't have any overrides (for any module ID) -> return myself
    get_hps_for_module(abc) -> "I have an override for abc -> return that"

lower level HPs objects (stored in self._per_module_overrides):
    User would NOT call this method, ever. You would always only call it on (top-level) self.hps instance.

sven1977 commented on rllib/core/learner/learner.py:
```diff
@@ -236,25 +274,17 @@ def __init__(
-            )
-
-        if module_spec is None and module is None:
+        if (module_spec is None) is (module is None):
```
simplified

sven1977 commented on rllib/core/learner/learner.py:
```diff
@@ -286,6 +316,15 @@ def __init__(
         # name includes the ModuleID as a prefix: optimizer_name=`[ModuleID]_[.. rest]`.
         self._module_optimizers: Dict[ModuleID, List[str]] = defaultdict(list)
 
+        self._user_configured_optimizers = is_overridden(
```
see above:
Users either:
- use RLlib's managed 1-optim per module setup.
- override `configure_optimizers_per_module`, but then have to do everything themselves (e.g. lr-scheduling, etc..).

sven1977 commented on rllib/core/learner/learner.py:
```diff
@@ -865,33 +1022,29 @@ def additional_update_for_module(
-        This includes grad clipping by value, norm, or global-norm, or other
-        algorithm specific gradient postprocessing steps.
+        # Handle lr-scheduling updates and apply new learning rates to the optimizers.
+        if not self._user_configured_optimizers:
```
We only auto-manage lr-scheduling iff user has NOT overridden `configure_optimizer_for_module` method.

sven1977 replied:
An alternative suggestion:
- Leave lr-updates inside additional-update method (it's NOT a gradient based, train batch-dependent "update" of the model)
- `Update` may or may not receive an additional timestep arg, but I don't think it's necessary as users should NOT override `update` anyways.
- Everything that the user wants to update every iteration anyways (and where there is no question that is should be updated every iteration), such as lr, entropy_coeff, kl, etc.. should NOT require a separate boolean arg in `addition_updates_for_module`'s signature. That seems overkill. If we argue against this, then we shouldn't even have a single `addition_update_for_module` method, but split it up into algorithm specific calls (similar to PPO's `update_kl` Policy mix-in methods :| ).
- For things like target net update, though, we should have it as target nets don't get updated every iteration.

kouroshHakha replied:
ok so this will be the TLDR of our conversations (The concrete convention)

The update and additional_update are assumed to be called on every iteration.
Scheduler stuff will get updated on every iterations. 
Things that may need a different frequency will have a dedicated flag to control what gets updated.

sven1977 commented on rllib/core/learner/tests/test_learner.py:
```diff
@@ -1,43 +1,24 @@
-from ray.rllib.core.rl_module.rl_module import SingleAgentRLModuleSpec
-from ray.rllib.core.testing.tf.bc_module import DiscreteBCTFModule
-from ray.rllib.core.testing.tf.bc_learner import BCTfLearner
+from ray.rllib.core.learner.learner import Learner
```
Unified with `test_torch_learner.py`.

sven1977 commented on rllib/core/learner/tf/tf_learner.py:
```diff
@@ -109,23 +108,7 @@ def compute_gradients(
-
-    @override(Learner)
-    def apply_gradients(self, gradients: ParamDict) -> None:
+    def apply_gradients(self, gradients: ParamDict):
```
postprocess_gradients moved to base class (now is framework agnostic).

kouroshHakha commented on rllib/core/learner/tests/test_learner.py:
```diff
@@ -77,57 +59,51 @@ def test_compute_gradients(self):
-            config.training(
-                grad_clip=0.75,
-                grad_clip_by="value",
+        """Tests the base grad clipping logic in `postprocess_gradients()`."""
```
I think you added this test in one of the previous PRs. The problem I see with this test is that it is not a unittest for learner as it has a lot of dependency on the correct implementation of learner and learner_group and how they interact with the config. We do not want that kind of dependency in this file which is mean to be only testing the basics of Learner itself. Let's modify this test to use `get_learner()` utility function?

kouroshHakha replied:
The problem right now is that if any thing else bugs out in the algorithm stack this test would fail.

kouroshHakha replied:
Also can we somehow avoid the repeatition of similar logic in this code. The test is overly complicated and verbose.

sven1977 replied:
Simplified the test case:
- no longer AlgorithmConfig dependent
- less convoluted b/c uses a simpler LearnerHPs object directly.

kouroshHakha commented on rllib/core/learner/tests/test_learner.py:
```diff
@@ -155,20 +131,26 @@ def test_postprocess_gradients(self):
             learner = learner_group._learner
+            # Pretend our computed gradients are our weights + 1.0.
+            grads = {
+                learner.get_param_ref(v): v + 1.0
```
We should probably just hard code the grads to 1.0 and -1.0 with prob of 50% each (and do not add value to it). If we do this we would be sure that all grad values are 1 and when you clip by value they would become 0.75 or -0.75

sven1977 replied:
I'll leave it for the reason of simplicity. If we want to hard-set grads to -1 or 1, we would have to do some shape-magic inside the for-loop (numpy, tf, torch?), whereas right now, the variable `v` takes care of that (framework-agnostic!).

kouroshHakha replied:
got it

kouroshHakha commented on rllib/core/learner/tests/test_learner.py:
```diff
@@ -225,109 +232,97 @@ def test_add_remove_module(self):
+    def _check_learner_states(self, framework, learner1, learner2):
         check(learner1.get_weights(), learner2.get_weights())
 
+        # Method to call on the localk optimizer object to get the optimizer's
```
typo: localk

sven1977 replied:
fixed

kouroshHakha commented on rllib/core/learner/tests/test_learner.py:
```diff
@@ -225,109 +232,97 @@ def test_add_remove_module(self):
 
         learner_1_optims_serialized = [
-            optim.get_config() for optim in learner1._optimizer_parameters.keys()
+            getattr(optim, method)() for optim in learner1._optimizer_parameters.keys()
         ]
         learner_2_optims_serialized = [
-            optim.get_config() for optim in learner2._optimizer_parameters.keys()
+            getattr(optim, method)() for optim in learner2._optimizer_parameters.keys()
```
These lines here, what are they doing that the previous lines are not doing? Should we get rid of them? one is checking the values in the list while the previous one is checking them as a dict. The dict one is more comprehensive.

sven1977 replied:
Not sure I understand. They check that the updates applied to an optimizer have the same effect, after you have restored the optimizer(s) from a checkpoint. This is not a new check I'm adding here, I'm just making this test framework-independent.

kouroshHakha commented on rllib/core/learner/tests/test_learner.py:
I think the tests need further clean up.

sven1977 replied:
done

kouroshHakha commented on rllib/core/rl_module/tests/test_marl_module.py:
```diff
@@ -14,13 +15,14 @@
 class TestMARLModule(unittest.TestCase):
     def test_from_config(self):
-
+        """Tests whether a MultiAgentRLModule can be constructed from a config."""
```
Can we create a a separate test unit for this? 

I want to leave this method as a simple unittest without dependency on AlgorithmConfig.

sven1977 replied:
Removed, this is now covered in TestAlgorithmConfig in a separate test.

kouroshHakha commented on rllib/core/testing/utils.py:
```diff
@@ -109,13 +111,20 @@ def get_learner(
-    return _cls(module_spec=spec, optimizer_config={"lr": learning_rate})
+    learner = _cls(
+        module_spec=spec,
+        learner_group_scaling_config=LearnerGroupScalingConfig(),
```
Did we need scaling config? It will get created in the constructor of Learner with default parameters if not specified as far as I remember.

kouroshHakha commented on rllib/utils/lambda_defaultdict.py:
```diff
@@ -0,0 +1,45 @@
+
+
+class LambdaDefaultDict(defaultdict):
+    """A defaultdict subclass that takes a lambda function as a default_factory.
```
What is the diff between this and the normal defaultdict? I feel like I am not clear so much on the diff.

sven1977 replied:
It's described one line below:
```
The lambda function is expected to accept a single argument: the missing key.

If a missing key is accessed, the provided lambda function is called with the
missing key as its argument. The returned value is stored in the dictionary
under that key and returned.
```
I'll add the description of why this is different (and important) wrt the standard defaultdict.

sven1977 replied:
Fixed description. Made it more clear what the difference is.

kouroshHakha replied:
oh independence of the default value based on key is the important key point. 👍

kouroshHakha commented on rllib/core/rl_module/rl_module.py:
```diff
@@ -59,13 +59,18 @@ class SingleAgentRLModuleSpec:
             observation space of an environment, would usually correspond to a
             one-hot encoded observation space of the RLModule because of preprocessing.
         action_space: The action space of the RLModule.
+        algorithm_config_overrides: An optional AlgorithmConfig override dict that
```
I don't think we should retrofit SingleAgentRLModuleSpecs to take bits of information that RLModule won't use directly. It's just not the right place. algorithm_config_overrides does not belong to SingleAgentRLModuleSpec
We only use this metric when creating a LearnerHyperParameters's override dict, right? So we can put this information in algorithm_config directly. For example we could do the following to pass in the algorithm_config_overrides.

```
config = config.multi_agent(algorithm_config_overrides)
```

sven1977 replied:
I was thinking about this, too. I think you are right. This information would sit at the wrong location.
It's actually the Learner that should hold this information on which Module receives which config override, not the module itself. I'll move this into the higher algo-config level directly, then derive the per-module settings within the Learner c'tor and keep things outside of Single|MultiAgentRLModuleSpec.

sven1977 replied:
done

kouroshHakha commented on rllib/algorithms/impala/impala_learner.py:
```diff
@@ -50,25 +54,35 @@ class ImpalaLearner(Learner):
         # Update entropy coefficient via our Scheduler.
-        new_entropy_coeff = self.entropy_coeff_scheduler.update(
-            module_id, timestep=timestep
+        new_entropy_coeff = self.entropy_coeff_schedulers_per_module[module_id].update(
```
I think this update should be tied to a input flag so that the user can control the frequency of each additional update event, independently from other kind of updates.

sven1977 replied:
good idea, will add

sven1977 replied:
Hmm, this doesn't make sense, imo. Then we should simply add different methods, one for each type of update, e.g. `update_entropy_coeff`, just like we had it in the old `PPOPolicies.update_kl()`.

I think that the PPO algorithm has to trust its framework-specific Learner subclasses to do the right thing (PPO-wise) in this one `additional_update` call.

Also, in this particular case (updating the entropy coeff), it really doesn't make sense to have more control. There is no reason to NOT update this coeff every single iteration.

I do see that this situation is different for say target net updates, where you don't want these updates to happen every iteration. So for these, we should indeed have a flag (rather than deciding inside additional_update by the `timestep` given as the algorithm should be the one deciding WHEN to do WHAT).

Bottom line: Yes, for target net updates, this makes sense. For lr/entropy/kl coeff updates not so much.

kouroshHakha reviewed (commented):
I didn't get to review the algorithm specific updates with details yet. But here are general comments for the rest of the stuff that I reviewed:

1. The unittest for learner has to be cleaned up and simplified. Some of the methods are not unittests and are actually integration tests with AlgorithmConfig and Algorithm. We don't want these kinds of dependencies.
2. I don't think the algorithm_override belongs to SingleAgentSpec object. Read my recommendation on where I think algorithm_override config belongs to. 
3. postprocess_gradients_for_module can be further simplified and made more moduler
4. I have one big concern about how we ignore the frequency at which different additional_update methods should take place. Right now, user doesn't have a control knob for that which is soon going to bit us back
5. I think AlgorithmConfig docstrings on the scheduler and optimizer configs should be made much more clear.
6. There are a lot of features that are not tested right now. For example the different formats in which we accept the lr_schedule / grad clip from algorithm_config is not tested. (Basically think about the features and different variations that were added in this PR, Are they all tested?)

kouroshHakha reviewed (commented):
I didn't get to review the algorithm specific updates with details yet. But here are general comments for the rest of the stuff that I reviewed:

1. The unittest for learner has to be cleaned up and simplified. Some of the methods are not unittests and are actually integration tests with AlgorithmConfig and Algorithm. We don't want these kinds of dependencies.
2. I don't think the algorithm_override belongs to SingleAgentSpec object. Read my recommendation on where I think algorithm_override config belongs to. 
3. postprocess_gradients_for_module can be further simplified and made more moduler
4. I have one big concern about how we ignore the frequency at which different additional_update methods should take place. Right now, user doesn't have a control knob for that which is soon going to bit us back
5. I think AlgorithmConfig docstrings on the scheduler and optimizer configs should be made much more clear.
6. There are a lot of features that are not tested right now. For example the different formats in which we accept the lr_schedule / grad clip from algorithm_config is not tested. (Basically think about the features and different variations that were added in this PR, Are they all tested?)

kouroshHakha commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -1652,24 +1651,40 @@ def training(
+                case, it is expected that your Learner subclass returns these exact
+                names and optimizers from the
+                `Learner.configure_optimizers_for_module()` method.
+            lr: The learning rate (float) or learning rate schedule in the format of
```
We should make this a TODO at the very least. But is there anything in the API stack makes this more complicated than it seems like today?

sven1977 replied:
I think we should not support anything like this from within the config alone. If users want to:
- use >1 optimizer per module
- use c'tor args other than learning rate for their optimizers
- use schedules other than piecewise
-> They must override configure_optimizer_for_module themselves and then handle everything (e.g. lr-scheduling) themselves.

kouroshHakha replied:
Understood. Makes sense.

sven1977 commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -1652,24 +1651,40 @@ def training(
             gamma: Float specifying the discount factor of the Markov Decision process.
-            lr: The default learning rate.
-            lr_schedule: Learning rate schedule. In the format of
+            optimizer_type: The type of the optimizer to use. RLlib will try to
```
Ah, Dict[str, str] is outdated. I'll remove.

sven1977 replied:
Actually, the more I think about it, the more I feel like we should really restrict the options the user has via the AlgorithmConfig (as opposed to by overriding `configure_optimizer_for_module`).
- only Adam allowed
- only lr (or a piecewise schedule) can be set
- only one optimizer per module possible
in any(!) other case, the user must override `configure_optimizer_per_module` and then handle all update/lr-scheduling logic themselves.

sven1977 replied:
Then we can remove the optimizer_type property again altogether.

sven1977 commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -897,7 +896,7 @@ def validate(self) -> None:
         # LR-schedule checking.
         if self._enable_learner_api:
-            Scheduler.validate(self.lr_schedule, "lr_schedule", "learning rate")
+            Scheduler.validate(self.lr, "lr", "learning rate")
```
Renamed args and enhanced docstring:
```
            setting_name: The property name of the schedule setting (within a config),
                e.g. `lr` or `entropy_coeff`.
            description: A full text description of the property that's being scheduled,
                e.g. `learning rate`.
```

sven1977 commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -1652,24 +1651,40 @@ def training(
+                individual learning rates (float) or learning rate schedules (see
+                above). In this case, it is expected that your Learner subclass returns
+                these exact names and optimizers from the
+                `Learner.configure_optimizers_for_module()` method.
```
Sorry, these were leftovers from before, when I had multiple optimizers per module in mind and making it possible to just configure those via the config, which we should not do anymore.

sven1977 commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -1652,24 +1651,40 @@ def training(
+                above). In this case, it is expected that your Learner subclass returns
+                these exact names and optimizers from the
+                `Learner.configure_optimizers_for_module()` method.
+            grad_clip: The value (float) to use for gradient clipping. Depending on the
```
Flipped and enhanced dosctrings to make things more clear.

sven1977 commented on rllib/core/learner/tf/tf_learner.py:
```diff
@@ -513,3 +471,20 @@ def _get_tensor_variable(self, value, dtype=None, trainable=False) -> "tf.Tensor
+    @staticmethod
+    @override(Learner)
+    def _get_clip_function() -> Callable:
+        from ray.rllib.utils.tf_utils import clip_gradients
```
Normally, I would agree (always import on top), but for clarity, I think that it makes this method easier to read. User goes to the method code and sees right away: "ah, we are returning the tf-specific clipping function imported from xyz here".

sven1977 commented on rllib/core/learner/torch/torch_learner.py:
```diff
@@ -75,14 +74,30 @@ def __init__(
-            torch.optim.Adam(self.get_parameters(module), lr=lr),
-        )
+
+        optimizers = {
+            "sgd": torch.optim.SGD,
+            "adam": torch.optim.Adam,
+            "adamw": torch.optim.AdamW,
+            "sparseadam": torch.optim.SparseAdam,
+            "adamax": torch.optim.Adamax,
+            "asgd": torch.optim.ASGD,
+            "lbfgs": torch.optim.LBFGS,
+            "rmsprop": torch.optim.RMSprop,
+            "rprop": torch.optim.Rprop,
+            "adagrad": torch.optim.Adagrad,
+            "adadelta": torch.optim.Adadelta,
+        }
```
Actually, the more I think about this, the more I'm against this entire `optimizer_type` setting. See my comment above on the restrictions we should put on users configuring their optimizer via the config alone (only Adam allowed, no c'tor args other than lr, no >1 optimizers per module).

sven1977 commented on rllib/core/learner/torch/torch_learner.py:
```diff
@@ -362,3 +346,16 @@ def _get_tensor_variable(
+    @staticmethod
+    @override(Learner)
+    def _get_clip_function() -> Callable:
+        from ray.rllib.utils.torch_utils import clip_gradients
```
see my explanation above

sven1977 commented on rllib/core/learner/learner.py:
```diff
@@ -111,12 +112,58 @@ class LearnerHyperparameters:
+    # Holds hyperparameters per module. This is not None only in the top-level
+    # Learner's self.hps (whose self.module is a `MARLModule`) and then contains the
+    # correct mappings from ModuleID to the derived LearnerHyperparameter objects.
+    # You can access a per-module HP sub-object by using the
```
simplified comment. Yeah, was a little convoluted. :)

sven1977 commented on rllib/core/learner/learner.py:
```diff
@@ -286,6 +323,15 @@ def __init__(
+        # Only manage optimizer's learning rate if user has NOT overridden
+        # the `configure_optimizers_for_module` method. Otherwise, leave responsibility
+        # to handle lr-updates entirely in user's hands.
+        if not self._user_configured_optimizers:
+            self._optimizer_lr_schedules: Dict[Optimizer, Scheduler] = {}
```
I think this is actually safer. If we screw up the `self._user_configured_optimizers`-logic we immediately would get a crash in our tests.

sven1977 commented on rllib/core/learner/learner.py:
```diff
@@ -466,9 +519,123 @@ def compute_gradients(
+            # method.
+            module_grads_dict = {}
+            for name in self._module_optimizers[module_id]:
+                optimizer = self._named_optimizers[name]
```
Hmm, yeah, I told you this would create trouble ;)

Let me think about this ...

sven1977 commented on rllib/core/learner/learner.py:
```diff
@@ -466,9 +519,123 @@ def compute_gradients(
+
+        # Loop through all optimizers of this `module_id`.
+        for name in self._module_optimizers[module_id]:
+            optim_name = name[
```
dude, compiling these names under the hood in this flattened format was NOT my idea to begin with. I had told you this was a recipe for disaster beforehand. You said there was no other way to structure this, so I went with it.

sven1977 commented:
On your points above:
```
1) The unittest for learner has to be cleaned up and simplified. Some of the methods are not unittests and are actually integration tests with AlgorithmConfig and Algorithm. We don't want these kinds of dependencies.
2) I don't think the algorithm_override belongs to SingleAgentSpec object. Read my recommendation on where I think algorithm_override config belongs to.
3) postprocess_gradients_for_module can be further simplified and made more moduler
4) I have one big concern about how we ignore the frequency at which different additional_update methods should take place. Right now, user doesn't have a control knob for that which is soon going to bit us back
5) I think AlgorithmConfig docstrings on the scheduler and optimizer configs should be made much more clear.
6) There are a lot of features that are not tested right now. For example the different formats in which we accept the lr_schedule / grad clip from algorithm_config is not tested. (Basically think about the features and different variations that were added in this PR, Are they all tested?)
```

1) done
2) agreed and fixed, moved out of SingleAgentRLModuleSpec into AlgorithmConfig directly
3) TODO
4) See my comments on this matter.
5) done
6) lr_schedule is properly validated by AlgorithmConfig.validate() (much more now than it used to before!). Not sure why we need extra formatting tests. Seems utter overkill given our extreme resource limitations to test any small detail.
Dude, `grad_clip` is a simple float! Are we validating floats now? There is a very detailed test case (which you just criticized as "too verbose") that I added and that tests any possible `grad_clip_by` setting in detail, including the resulting clipped gradients.
